src/hotspot/share/gc/shared/collectedHeap.hpp

```diff
@@ -518,6 +521,20 @@ class CollectedHeap : public CHeapObj<mtGC> {
   virtual HeapWord* allocate_loaded_archive_space(size_t size) { return nullptr; }
   virtual void complete_loaded_archive_space(MemRegion archive_space) { }
 
+  // Commit the heap memory for CDS heap archive area according to the requested size.
+  // Preferred address is treated as a hint for the location of the archive area in the heap.
+  // The returned memory range may or may start at the preferred address.
+  // Return MemRegion corresponding to the commited heap memory.
+  virtual HeapWord* alloc_archive_space(size_t word_size, HeapWord* preferred_addr) { return nullptr; }
+
+  // This must be called after alloc_archive_space, and after class loading has occurred.
+  // GC policy can take necessary actions to make the archive space usable,
+  // such as inserting filler objects to make it parseable, or updating block offset table.
+  virtual void fixup_archive_space(MemRegion range) { return; }
+
+  // Handle any failure in loading the CDS archive area.
+  virtual void handle_archive_space_failure(MemRegion range) { return; }
```

Comment:
For both `fixup_archive_space` and `handle_archive_space_failure`: Why do we need to pass in the MemRegion? The only valid argument in both cases is the MemRegion that had been previously established by a call to `alloc_archive_space`. 

Why not let CollectedHeap remember the MemRegion established by `alloc_archive_space`, thereby also guaranteeing that `alloc_archive_space` can only be called once, and refer to that in the followup calls?

Reply:
I don't see any reason why we can't do that. Only reason could be that we need the MemRegion in FileMapInfo as well, so we end up duplicating the MemRegion in CollectedHeap and FileMapInfo. 
But looking at the uses of `_mapped_heap_memregion` in FileMapInfo.cpp I think cds only needs the start address of the archive space in the heap. Size of the region must be same as the used bytes of `MetaspaceShared::hp` region. So we can ideally replace `_mapped_heap_memregion` with something like`_archive_space_load_address`.
@iklam @tschatzl what is your opinion?